lib/posix-process/brk.c

```diff
@@ -0,0 +1,268 @@
+			 * out starting from cur_brk to the start of the next
+			 * page.
+			 */
+			zlen = PAGE_ALIGN_UP(cur_brk) - cur_brk;
+		} else {
+			/* If addr is within the page cur_brk is in then zero
+			 * out starting from cur_brk to addr.
+			 */
+			zlen = addr - cur_brk;
+		}
+
+		memset((void *)cur_brk, 0, zlen);
+
+		return 0;
+	}
+
+	/*
+	 * Decrement: Free up whole pages between previous cur_brk and addr.
+	 */
+
+	/* Align boundaries to identify full pages */
+	free_start = PAGE_ALIGN_UP(addr);
+	free_end = PAGE_ALIGN_UP(cur_brk);
+
+	/* Check if we have at least one full page to free */
+	if (!(free_start < free_end))
+		return 0;
+
+	uk_pr_debug("Freeing up brk range 0x%lx-0x%lx (%lu pages)\n",
+		    free_start, free_end,
+		    (free_end - free_start) >> __PAGE_SHIFT);
+
+	rc = uk_vma_advise(uk_vas_get_active(), free_start,
+			   free_end - free_start, UK_VMA_ADV_DONTNEED, 0);
+	if (unlikely(rc)) {
+		uk_pr_err("Failed to free up pages between 0x%lx-0x%lx\n",
+			  free_start, free_end);
+		return rc;
+	}
+
+	return 0;
+}
+#else /* !CONFIG_LIBUKVMEM */
+static inline int do_brk_inrange(__uptr cur_brk, __uptr addr)
+{
+	/*
+	 * Increment: We must zero out the bumped region as some applications
+	 * might expect.
+	 *
+	 * See glibc's MORECORE_CLEARS. We go with MORECORE_CLEARS == 2 to be
+	 * on the safe side.
+	 */
+	if (addr > cur_brk) {
+		uk_pr_debug("Zeroing 0x%lx-0x%lx...\n", cur_brk, addr);
+		memset((void *)cur_brk, 0x0, addr - cur_brk);
+	}
+
+	/*
+	 * Decrement: Do nothing as we are in one single allocated region
+	 * whose individual pages we cannot arbitrarily free.
+	 */
+
+	return 0;
+}
+#endif /* !CONFIG_LIBUKVMEM */
+
+UK_LLSYSCALL_R_DEFINE(void *, brk, void *, addr)
+{
+	struct posix_process *pprocess;
+	__uptr base, cur_brk, addr_arg;
+	int rc;
+
+	pprocess = uk_pprocess_current();
+	UK_ASSERT(pprocess);
+
+	base = (__uptr)pprocess->brk_ctx.base;
+	cur_brk = base + pprocess->brk_ctx.pos;
+	addr_arg = (__uptr)addr;
+
+	if (!IN_RANGE(addr_arg, base, HEAP_LEN)) {
+		uk_pr_debug("0x%lx outside PID %d's brk range 0x%lx-0x%lx, keep brk @ 0x%lx\n",
+			    addr_arg, pprocess->pid,
+			    base, base + HEAP_LEN, cur_brk);
+		return (void *)cur_brk;
+	}
+
+	rc = do_brk_inrange(cur_brk, addr_arg);
+	if (unlikely(rc))
+		return NULL;
+
+	uk_pr_debug("PID %d brk @ 0x%lx (brk heap region: 0x%lx-0x%lx)\n",
+		    pprocess->pid, addr_arg, base, base + HEAP_LEN);
+
+	pprocess->brk_ctx.pos = addr_arg - base;
+
+	return addr;
+}
+
+#if UK_LIBC_SYSCALLS
+int brk(void *addr)
+{
+	long rc;
+
+	rc = uk_syscall_r_brk((long)addr);
+	if (unlikely(!rc)) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	return 0;
+}
+
+#if CONFIG_LIBNOLIBC
+void *sbrk(intptr_t inc)
+{
+	struct posix_process *pprocess;
+	long req_base, rc;
+	void *prev_base;
+
+	pprocess = uk_pprocess_current();
+	UK_ASSERT(pprocess);
+
+	/* We are increasing or reducing our range */
+	prev_base = (__u8 *)pprocess->brk_ctx.base + pprocess->brk_ctx.pos;
```

Comment:
Nit: Although casting to `__u8 *` is valid, the implementation of the `brk()` syscall casts to `uptr_t` for arithmetic. It would be nice if there was some consistency across the file.